### PR TITLE
be more robust about checking embedding modules for lora finetunes

### PR DESCRIPTION
**Pull request by winglian**
@mtenenholtz brought up in the discord that phi has a more nuanced embedding module name. this PR attempts to handle other architectures a little more gracefully

### CODE REVIEW

File: src/axolotl/utils/models.py [hunk 1]
```diff
@@ -52,6 +56,20 @@ def check_model_config(cfg: DictDefault, model_config: AutoConfig):
...
+    lora_modules_to_save = get_linear_embedding_layers(model_config.model_type)
+    if (
+        cfg.adapter
+        and cfg.tokens
+        and (
+            not cfg.lora_modules_to_save
+            or not all(x in cfg.lora_modules_to_save for x in lora_modules_to_save)
+        )
+    ):
+        lora_modules_to_save = ", ".join(map(lambda x: f"`{x}`", lora_modules_to_save))
```
Comment (NanoCode012) [inline src/axolotl/utils/models.py:68]:
```suggestion
        lora_modules_to_save = ", ".join([f"`{x}`" for x in lora_modules_to_save])
```

Could this be refactored to so?

File: src/axolotl/utils/models.py [hunk 2]
```diff
@@ -149,11 +168,9 @@ def load_tokenizer(cfg):
...
                     or not all(
-                        x in cfg.lora_modules_to_save
-                        for x in ["embed_tokens", "lm_head"]
+                        x in cfg.lora_modules_to_save for x in lora_modules_to_save
                     )
                 )
-                and (model_config.model_type in ["llama", "mistral", "mixtral"])
             ):
                 raise ValueError(
                     "Please set lora_modules_to_save to ['embed_tokens', 'lm_head'] when using an adapter and changing the special tokens."
```
Comment (NanoCode012) [inline src/axolotl/utils/models.py]: You missed the string here also

File: tests/test_validation.py [hunk 3]
```diff
@@ -717,10 +731,48 @@ def test_add_tokens_adapter(self):
...
         )
 
-        validate_config(cfg)
+        check_model_config(cfg, model_config)
+
+    def test_phi2_add_tokens_adapter(self):
+        cfg = DictDefault(
+            {"adapter": "qlora", "load_in_4bit": True, "tokens": ["<|imstart|>"]}
+        )
+        model_config = DictDefault({"model_type": "phi-msft"})
```
Comment (NanoCode012) [inline tests/test_validation.py:745]: Would it be possible to load the actual config from their repo or is that not good for unit tests?

Comment (winglian) [reply tests/test_validation.py:745]: for the model config? I suppose we could, but that would require fleshing out the `cfg` variable more. This is like the minimum required to test the function and is easy to grok.

Comment (NanoCode012) [reply tests/test_validation.py:745]: Ok, sounds good

**Pull request merged**